### ISSUE TITLE
Integration with OS Keychains

### ISSUE THREAD

**Issue by CMCDragonkai** [MatrixAI/Polykey#201]
### Is your feature request related to a problem? Please describe.

Applications often delegate credential storage to a third party program on the OS rather than writing it themselves. Since Polykey is not a well-known credential storage, then PK won't immediately be used as a credential storage system. Therefore this limits the UX of using Polykey. It's an integration problem.

Browsers maintain their own credentials, and every other system has their own as well.

These third party programs include:

* D-Bus Secret Service
* Apple macOS keychain
* Microsoft Windows Credential Manager (`wincred`)
* pass (unix CLI tool)
* https://lwn.net/Articles/490518/

For example Docker is capable of integrating into these keychains as you can see here: https://github.com/docker/docker-credential-helpers

> By default, Docker looks for the native binary on each of the platforms, i.e. “osxkeychain” on macOS, “wincred” on windows, and “pass” on Linux. A special case is that on Linux, Docker will fall back to the “secretservice” binary if it cannot find the “pass” binary. If none of these binaries are present, it stores the credentials (i.e. password) in base64 encoding in the config files described above.
> https://docs.docker.com/engine/reference/commandline/login/#credentials-store

This shows you that Docker login credentials are not even encrypted at rest.

Android and iOS are also important here. In fact Android and iOS integration I believe actually supersedes browser integration. So you don't really create a "browser plugin" for the mobile phone browsers, but a password manager can be integrated to the OS. For example: https://developer.android.com/guide/topics/text/autofill and https://developer.apple.com/documentation/security/password_autofill/

### Describe the solution you'd like

It would be nice for Polykey to capable of performing secret injection into the OS keychain to use these as potential integration points.

### Describe alternatives you've considered

A generic integration point on desktop OSes is using the clipboard or performing autofill.

For autofill/autotype on desktop, we should look at this:

* https://keepass.info/help/base/autotype.html
* https://keepass.info/help/v2/autotype_obfuscation.html
* https://community.bitwarden.com/t/auto-type-autofill-for-logging-into-other-desktop-apps/158/106 - see the last few comments describing some of this

I think desktops don't have a direct autofill/autotype feature builtin unlike mobile systems. But this would be quite useful for desktop applications.

### Additional context

* https://gitlab.com/MatrixAI/Engineering/Polykey/Polykey/-/issues/6
* https://www.freedesktop.org/wiki/Specifications/secret-storage-spec/ - Secret Storage Specification for X Desktop (can then integrate to Linux OS), also:
  - https://rtfm.co.ua/en/what-is-linux-keyring-gnome-keyring-secret-service-and-d-bus/
  - https://github.com/keepassxreboot/keepassxc/issues/6274

**Comment by CMCDragonkai** [comment]
This is new feature request, and it requires review into our product vision, and if accepted, we are meant to create a design issue for this.

**Comment by CMCDragonkai** [comment]
Some recent experiments on running a grafana-agent due to #599 was interesting in showcasing how systemd works to manage credentials especially on NixOS.

This has some relation with #169.

The grafana-agent service config looks like this in NixOS:

```ts
{
  enable = true;
  settings = {
    integrations = {
      prometheus_remote_write = [
        {
          basic_auth = {
            username = "\${METRICS_USERNAME}";
            password = "\${METRICS_PASSWORD}";
          };
          url = "\${METRICS_URL}";
        }
      ];
    };
    logs = {
      configs = [
        {
          clients = [
            {
              basic_auth = {
                username = "\${LOGS_USERNAME}";
                password = "\${LOGS_PASSWORD}";
              };
              url = "\${LOGS_URL}";
            }
          ];
        }
      ];
    };
    metrics = {
      configs = [
        {
          name = "integrations";
          remote_write = [
            {
              basic_auth = {
                username = "\${METRICS_USERNAME}";
                password = "\${METRICS_PASSWORD}";
              };
              url = "\${METRICS_URL}";
            }
          ];
        }
      ];
      wal_directory = "\${STATE_DIRECTORY}";
    };
  };
  credentials = {
    METRICS_URL = "/run/keys/grafana_agent/metrics_url";
    METRICS_USERNAME = "/run/keys/grafana_agent/metrics_username";
    METRICS_PASSWORD = "/run/keys/grafana_agent/metrics_password";
    LOGS_URL = "/run/keys/grafana_agent/logs_url";
    LOGS_USERNAME = "/run/keys/grafana_agent/logs_username";
    LOGS_PASSWORD = "/run/keys/grafana_agent/logs_password";
  };
}
```

What's interesting here is that the `credentials` specially managed by the Nix derivation for the systemd unit.

Here you can see that they are key values where the value is a path to a file. The file path is in `/run/keys` which is supposed to be owned by the root user. By keeping as files instead of directly embedding them in the config, we avoid loading the credentials directly into the evaluated Nix which would be globally readable in the `/nix/store`.

This is then used here: https://github.com/NixOS/nixpkgs/blob/3dc440faeee9e889fe2d1b4d25ad0f430d449356/nixos/modules/services/monitoring/grafana-agent.nix#L127C1-L161C7

The 2 key things are:

```
LoadCredential = lib.mapAttrsToList (key: value: "${key}:${value}") cfg.credentials;
```

Here you can see how Nix is used as a templating language for arbitrary configuration. This systemd directive basically reads the file paths and puts them into a special location called the `CREDENTIALS_DIRECTORY`.

This is all documented here: https://systemd.io/CREDENTIALS/

Later inside the script that actually gets executed, it does this:

```sh
# Load all credentials into env if they are in UPPER_SNAKE form.
if [[ -n "${CREDENTIALS_DIRECTORY:-}" ]]; then
  for file in "$CREDENTIALS_DIRECTORY"/*; do
    key=$(basename "$file")
    if [[ $key =~ ^[A-Z0-9_]+$ ]]; then
      echo "Environ $key"
      export "$key=$(< "$file")"
    fi
  done
fi
```

Thus ensuring that all the files under `CREDENTIALS_DIRECTORY` gets loaded as an environment variable.

By default systemd doesn't do this, because environment variables are not as isolated as files as per:

> Access to credentials is restricted to the service’s user. Unlike environment variables the credential data is not propagated down the process tree. Instead each time a credential is accessed an access check is enforced by the kernel. If the service is using file system namespacing the loaded credential data is invisible to all other services.

Basically this is the problem of the confused deputy. If there other processes whether they are sibling processes or child processes with the same ambient environment, they may be able to access those credentials by looking at the environment of a given process. File systems then have more fine grained access control that can be applied to a given situation. But with script it just gives some flexibility in using environment variables.

---

Note that I have some problem with it reloading credentials: https://github.com/systemd/systemd/issues/21099

This ends up requiring a restart of the service atm.

**Comment by CMCDragonkai** [comment]
The PAM framework on Linux allows one to automatically unlock the OS keychain using the same password that is used during login of the user session. This allows one to open the keychain (containing various necessary secrets) and allow the keychain to start injection operations.

In this sense, the usage of a keychain here is an upgrade from sprawling plaintext secrets over the home directory. The OS user-session system gates access to the home directory normally, and most programs still expect the ability to load secrets from a file path. An OS keychain like PK at the very least enables security-at-rest compared to plaintext secrets in the home directory.

In this sense it can help solve the secret zero problem of user-level background service of PK by pushing the buck to the OS login session.

**Comment by CMCDragonkai** [comment]
https://github.com/MatrixAI/nixpkgs-matrix-private/pull/16#issuecomment-2670456472 - experiment with gnome keyring in our nixos platforms.

What seahorse looks like with `nix shell 'nixpkgs#seahorse'`.

![Image](https://github.com/user-attachments/assets/3559665d-af63-4e20-9e14-7d1d3056ee82)

**Comment by CMCDragonkai** [comment]
Related issues:

* https://github.com/MatrixAI/nixpkgs-matrix-private/issues/21
* https://github.com/matrixai/orchestrator/issues/11

Related projects:

* https://www.funtoo.org/Funtoo:Keychain (not sure how this plays into the UI like seahorse/gnome keyring).

On a separate note gpg-agent does not require a root password (and thus is not integrated with PAM):

You're correct in observing that `gpg-agent` is typically launched automatically during a user's login session. Its primary function is to manage and cache the passphrases associated with your GPG keys, facilitating cryptographic operations such as signing and decryption.

**Authentication and Passphrase Management:**

- **Agent Initialization:** Starting `gpg-agent` does not require entering a passphrase. The agent itself is a background process responsible for handling cryptographic operations and caching passphrases.

- **Passphrase Prompting:** When you perform an operation that requires access to a private key (e.g., signing a commit or decrypting a message), `gpg-agent` invokes a `pinentry` program to prompt you for the passphrase associated with that specific private key. This ensures that even though the agent is running, access to your private keys remains protected by their respective passphrases.

**Passphrase Caching Behavior:**

- **Default Cache Duration:** By default, `gpg-agent` caches your passphrase for 600 seconds (10 minutes) after the last use. This means that if you perform another operation requiring the same passphrase within this period, you won't be prompted again.

- **Maximum Cache Duration:** The maximum time a passphrase can be cached is 2 hours (7200 seconds) by default. After this period, you'll be prompted to re-enter the passphrase, regardless of activity.

**Configuring Cache Settings:**

You can adjust these caching durations by modifying the `gpg-agent.conf` file:

1. **Edit the Configuration File:**
   ```bash
   nano ~/.gnupg/gpg-agent.conf
   ```


2. **Add or Modify the Following Lines:**
   ```bash
   default-cache-ttl 3600        # Cache duration in seconds after last use (e.g., 1 hour)
   max-cache-ttl 14400           # Maximum cache duration in seconds (e.g., 4 hours)
   ```


3. **Apply the Changes:**
   ```bash
   gpgconf --reload gpg-agent
   ```


This configuration ensures that your passphrase is cached for 1 hour after the last use, with a maximum cache duration of 4 hours.

**Security Considerations:**

- **Session-Based Caching:** The cache durations are reset with each use during an active session. However, if you log out or restart your system, the cached passphrases are cleared, and you'll need to re-enter them upon the next use.

- **Manual Cache Clearing:** If you wish to clear the cached passphrases manually, you can restart `gpg-agent` or use the appropriate commands to flush the cache.

In summary, while `gpg-agent` manages the caching of passphrases to enhance usability, it does not compromise the security of your private keys. Each cryptographic operation requiring access to a private key will prompt for the associated passphrase unless it is currently cached, ensuring that your keys remain protected.

---

I actually kind of dislike GPG's UI. It would be ideal that signing just happens entirely with Polykey, and I dispense with git's integration into GPG. But given that git has native integration with GPG, that would imply that PK can facilitate commit signing similar to GPG (with a GPG-like UI). If we do this, I would like this to be solved simultaneously: https://github.com/NixOS/nixpkgs/issues/57779

**Comment by CMCDragonkai** [comment]
This comment delves into the specifics of integrating the *user-level* Polykey service with PAM, drawing on the investigation documented in `notes-on-provisioning-users.md` (approx. lines 176-234), and relating it to the goals outlined in this issue's specification.

### Context: User vs. System Service Secrets

As the issue specification clarifies, we have two distinct scenarios for providing the "secret zero" (password or recovery code) to Polykey agents:

1.  **System-Level Agent:** Represents the machine's identity. The proposed solution using TPM2 and `systemd-creds` (`LoadCredentialEncrypted=`) appears robust for securely injecting secrets without needing the (often locked) root password after initial setup. This leverages hardware security and standard systemd mechanisms.
2.  **User-Level Agent:** Represents the user's identity/gestalt. Currently relies on `~/.polykeypass`. The goal is to eliminate this file and instead leverage the user's *login password* via PAM integration, mimicking the seamless experience of OS keychains like macOS Keychain or GNOME Keyring.

### PAM Exploration Insights (`notes-on-provisioning-users.md`)

Our investigation into PAM yielded several relevant points:

1.  **PAM's Role:** PAM (Pluggable Authentication Modules) primarily governs authentication, account management, session setup/teardown, and password changes for *interactive* user sessions and services (e.g., `login`, `sshd`, `sudo`, `passwd`). It uses service-specific configurations (`/etc/pam.d/*`) that invoke specific PAM modules (`.so` files).

2.  **Module Function:** Modules are shared libraries implementing different checks or actions (e.g., `pam_unix.so` checks `/etc/shadow`, `pam_fprintd.so` communicates with the fingerprint daemon). They execute sequentially based on the configuration.

3.  **GNOME Keyring Example:** The integration of `gnome-keyring` is particularly instructive. The line `session optional pam_gnome_keyring.so auto_start` in PAM configurations (like `/etc/pam.d/login` or `/etc/pam.d/sddm`) suggests that the PAM `session` management phase can trigger the startup of the keyring daemon. Crucially, `pam_gnome_keyring.so` also likely handles unlocking the user's default keyring using the password successfully authenticated earlier in the PAM stack (`auth` phase). Once unlocked, the running `gnome-keyring-daemon` acts as an *egress point* for secrets, accessible by other user applications (e.g., Seahorse, Evolution, `git credential-gnome-keyring`) via IPC (likely DBus).

4.  **Polykey *Using* PAM (Goal):** The objective isn't necessarily for Polykey to *be* a PAM module that *other* services use for authentication (which would require a `pam_polykey.so`). Instead, the Polykey *user-level service* needs to *hook into the user's login process via PAM* to securely receive the login password and use it to decrypt its own agent state/vault, thereby eliminating `~/.polykeypass`.

5.  **Implementation Challenge:** How can the Polykey user service (a Node.js application) interact with the inherently C-based PAM stack during login?
    *   **Custom PAM Module:** The most direct way might be to create a minimal `pam_polykey.so` module (likely in C or Rust). This module would run during the PAM `auth` or `session` phase. Upon successful authentication upstream (e.g., by `pam_unix.so`), our module could capture the password from the PAM state (using functions like `pam_get_item(..., PAM_AUTHTOK, ...)` if available) and securely transmit it to the waiting Polykey user agent process via IPC (DBus, Unix socket). This requires careful implementation to handle security and process lifecycle.
    *   **Using Existing Modules:** Alternatively, investigate modules like `pam_exec.so` or `pam_script.so`. These allow running external commands during PAM phases. However, securely passing the password via stdin or environment variables is often problematic and generally discouraged. It's unclear if they provide a sufficiently secure channel for this purpose.
    *   **Post-Login Integration:** Could the Polykey user service start *after* login and query another component that *was* unlocked by PAM? For example, if `gnome-keyring` is present and unlocked, could Polykey request a derived key or use it as a backend? This creates dependencies on other keyrings.

6.  **Complexity:** Building and correctly integrating a custom PAM module is non-trivial and requires careful handling of the PAM API, security contexts, and potential interactions with other modules in the stack.

### Connecting PAM to Issue Tasks

This PAM exploration directly informs **Task 2: Rewrite the system level nixos module service and user level service...**. The user-level service rewrite must incorporate the chosen PAM integration mechanism. It also impacts **Task 4/5: Test ... Ensure persistent Polykey agent creation**, as seamless unlocking via PAM is critical for the user agent's usability and effective persistence across logins.

### Summary for User-Level Service

Achieving the desired PAM integration for the user-level service likely requires creating a dedicated, minimal PAM module (`pam_polykey.so`) that:

1.  Integrates into the PAM stack (e.g., in `/etc/pam.d/system-login` or display manager PAM config like `/etc/pam.d/sddm`).
2.  Runs after successful primary authentication (`auth` phase sufficient).
3.  Securely retrieves the authenticated password from the PAM context.
4.  Communicates this password via secure IPC to the Polykey user agent, allowing it to unlock its state.

This mirrors the `gnome-keyring` model and provides the target user experience, distinct from the TPM-based approach for the system-level agent. Further research into the PAM development API (specifically obtaining `PAM_AUTHTOK` securely) and secure IPC methods between a C/Rust module and a Node.js process is needed.

**Comment by CMCDragonkai** [comment]
Okay, here is an elaboration on OS keychains, focusing on `gnome-keyring` and `seahorse` based on your exploration (`notes-on-provisioning-users.md`, approx. lines 190-276), and contrasting them with Polykey's approach. V=5.

### OS Keychains: The Concept

Operating system keychains (like macOS Keychain Access, Windows Credential Manager, or Linux keyring systems) serve as centralized, secure databases for storing sensitive user credentials. Their primary goal is to consolidate secrets such as:

*   Website and application passwords
*   Private cryptographic keys (SSH, GPG) and their passphrases
*   Digital certificates (TLS/SSL)
*   Secure notes

Instead of applications storing these secrets insecurely in configuration files or repeatedly prompting the user, they can leverage the OS keychain's API. Typically, the keychain itself is encrypted using a master password, often derived from or directly linked to the user's login password. This allows for a form of single sign-on (SSO) within the OS session – once the user logs in, authenticated applications can retrieve necessary credentials from the unlocked keychain without further prompts.

### `gnome-keyring` and `seahorse`: A Linux Implementation

Your exploration using `seahorse` provided a practical look into `gnome-keyring`, a common keychain implementation on Linux desktops, particularly GNOME.

1.  **`gnome-keyring` (The Backend Daemon):**
    *   **Function:** This is the core service (`gnome-keyring-daemon`) that runs in the background. It manages encrypted files, typically stored in `~/.local/share/keyrings/`. Each `.keyring` file is a vault, potentially holding multiple secrets (items).
    *   **Encryption & Unlocking:** As noted (lines 265-270), keyrings can be password-protected. The crucial "login" keyring is typically unlocked automatically when the user logs into their graphical session. This seamless unlocking is achieved via PAM integration (specifically `pam_gnome_keyring.so`, lines 223-228). During the login process (handled by PAM), after successful authentication (e.g., via password check with `pam_unix.so`), the `pam_gnome_keyring.so` module uses the authenticated password to unlock the user's default keyring.
    *   **API Access:** The daemon provides access to stored secrets via a standard DBus interface (the Freedesktop Secret Service API). Applications use libraries like `libsecret` to interact with this API, allowing them to store and retrieve credentials programmatically.

2.  **`seahorse` (The Frontend GUI):**
    *   **Function:** This application acts as a graphical user interface *for* `gnome-keyring` (and potentially other backends supporting the Secret Service API). It allows users to view and manage the secrets stored by the backend daemon.
    *   **Observed Capabilities (lines 229-264):** Your interaction with `seahorse` revealed its scope:
        *   **Password Management:** Creating, viewing, and deleting basic password items stored directly within `gnome-keyring`'s `.keyring` vaults. It provides basic name/secret pairs.
        *   **GPG Key Management:** It discovers and displays GPG keys found in the user's `~/.gnupg` directory, showing IDs, fingerprints, subkeys, and associated User IDs (name/email). It allows changing passphrases and exporting keys. It also integrates with GPG keyservers (like `keys.openpgp.org`) for searching and "syncing" public keys – a centralized/federated model for key discovery (line 239-241). It acts as a frontend viewer/metadata manager, not a replacement for `gpg` or `gpg-agent` for performing cryptographic operations.
        *   **SSH Key Management:** Similarly, it discovers SSH keys (likely from `~/.ssh/`) and potentially `authorized_keys`, displaying metadata and allowing passphrase changes or new key generation. Again, it manages *storage* aspects, not SSH protocol operations.
        *   **Certificate Management:** It displays system-wide trusted TLS CA certificates (e.g., from `/etc/ssl/certs/ca-certificates.crt`, mentioned lines 245-253) and potentially user-added certificates. It typically allows viewing details and managing trust settings but might have limitations on creating private keys/certs, especially in read-only Nix store environments (line 268).
    *   **Limitations Noted:** You observed that `seahorse`/`gnome-keyring` primarily focuses on *managing* existing keys/secrets and their metadata (especially passphrases) for legacy systems (GPG, SSH, TLS PKI). It doesn't perform the core signing/encryption operations itself (line 236). The password vault feature is quite basic compared to dedicated password managers. Its integration feels like a convenient UI layer over disparate, older systems rather than a unified, modern cryptographic system. You also noted potential issues with graphical prompts in non-graphical sessions (line 237).

### How Polykey (PK) Differs and Evolves

Polykey represents a fundamentally different approach compared to traditional OS keychains like `gnome-keyring`. The key distinctions lie in architecture, scope, and capabilities:

1.  **Scope: Identity & Authority vs. Secret Storage:**
    *   `gnome-keyring`: Primarily a secure *storage container* for secrets local to an OS user session, with helper functions for managing passphrases of external crypto systems (GPG/SSH). Its core value is centralized local storage and PAM-based unlock integration.
    *   `Polykey`: A comprehensive system for *distributed cryptographic identity and authority management*. Secret storage (in vaults) is just one component. Its core value lies in establishing verifiable identities (sigchains), managing fine-grained access control policies, enabling secure peer-to-peer communication, and performing cryptographic operations directly.

2.  **Architecture: Local vs. Distributed:**
    *   `gnome-keyring`: Inherently local to the machine and user session. Data is stored in local files.
    *   `Polykey`: Designed for distributed operation. Agents (nodes) can exist on multiple machines, forming gestalts (representing a user or service). Vaults and sigchains can be replicated and synchronized across these nodes according to defined policies.

3.  **Active Crypto Engine vs. Passive Store:**
    *   `gnome-keyring`: Passively stores keys/passphrases for other tools (like `gpg` or `ssh`) to use. `seahorse` mainly manages metadata.
    *   `Polykey`: Actively *uses* its managed keys for signing, verification, encryption, decryption, authentication, and authorization within its own protocol framework. It *is* the cryptographic engine for its ecosystem.

4.  **Integration Model:**
    *   `gnome-keyring`: Integrates deeply with the OS login via PAM and with desktop applications via the standard Secret Service DBus API.
    *   `Polykey`: Integration occurs primarily via its own CLI, APIs, and SDKs. While PAM integration is being considered for the *user-level* agent's unlock mechanism (to replace `~/.polykeypass`), its core operation is less tied to OS-specific session management. The *system-level* agent uses systemd features (like TPM integration).

5.  **Protocol Evolution:**
    *   `gnome-keyring`: Provides a bridge to manage secrets for established, often legacy protocols (GPG keyservers using HKPS/LDAP, SSH keys, X.509 TLS certs).
    *   `Polykey`: Defines its own modern, cryptographically robust protocols for identity (sigchains), synchronization (node graph/vault sync), communication, and access control, drawing inspiration but evolving beyond systems like GPG, Keybase, and blockchain identity models (line 235-238). It aims to supersede, not just manage, older systems where appropriate.

6.  **Key Management Paradigm:**
    *   `gnome-keyring`/`seahorse`: Exposes interfaces to manage keys primarily defined by external systems (GPG, SSH files).
    *   `Polykey`: Manages its own cryptographic keys directly within its secure storage (vaults), tying them intrinsically to the agent's identity and using them for its core functions.

In essence, while `gnome-keyring`/`seahorse` provides a convenient way to manage secrets for existing tools within a Linux desktop session, leveraging PAM for seamless unlocking, Polykey aims to be a foundational layer for distributed identity, authority, and secure operations, offering a more powerful, flexible, and modern cryptographic framework that extends far beyond local secret storage. The PAM integration contemplated for Polykey's user agent specifically aims to replicate the seamless *unlocking* experience of `gnome-keyring` but within Polykey's fundamentally different architectural context.

**Comment by CMCDragonkai** [comment]
Example of desktop apps that try to "detect" the OS keyring and use it:

![Image](https://github.com/user-attachments/assets/0e2c0fc0-efc3-4975-ae37-433d2a324197)

However even though my gnome-keyring is in fact running, it doesn't detect it, uses "weaker encryption" - whatever that means - and probably means it ends up storing it in the HOME directory somewhere.

Therefore PK facilitating this means that other "desktop" apps following protocol will end up using PK to store secrets, just like how password apps are like for web stuff.
